core/src/main/java/org/apache/accumulo/core/util/ratelimit/SharedRateLimiterFactory.java

```diff
@@ -112,9 +112,9 @@ protected void update() {
     }
     for (Map.Entry<String,WeakReference<SharedRateLimiter>> entry : limitersCopy.entrySet()) {
       try {
-        if (Objects.nonNull(entry.getValue().get())) {
-          entry.getValue().get().update();
-        }
+        Objects.requireNonNull(entry.getValue().get()).update();
+      } catch (NullPointerException npex) {
```

Comment:
These changes were added to avoid spotbugs throwing  a "Possible null pointer dereference" error. It seems like adding the extra catch block for NPE is redundant and a single catch for `Exception` should be sufficient, however without it spotbugs will throw the mentioned error.

I am wondering if it would be best to suppress spotbugs in this case and let the singular `Exception` catch block handle the NPE. I am open to suggestions on this.

Reply:
Spotbugs is signaling that WeakReferences stored in the map may actually be a reference to null, if the referenced object had been garbage collected, but not yet removed from the map. It would be better to avoid inserting any weak references into this copy, since the whole point is to get strong references to call update on them. It'd probably be enough to simply have an `if` statement to check to see if they are null before calling `update` on them after the copy.

Alternatively, you could map the values by calling `WeakReference.get()` on the values, and filtering out the null ones. And, since we don't use the keys, it's even easier. Something like:

```java
  List<SharedRateLimiter> limitersCopy = new ArrayList<>();
  synchronized(activeLimiters) {
    limitersCopy = activeLImiters.values().stream().filter(Objects::nonNull).collect(Collectors.toList());
  }
  limitersCopy.forEach(SharedRateLimiter::update);
```

Of course, it might not be exactly that simple, because I just realized we *do* use the keys for the log message if there's an exception. But hopefully you get the point... filter out or check for nulls from the copy before calling update, because WeakReference.get() can be null.